File: packages/server/src/room.ts
import { WebSocket } from 'ws';
import { Game, PHYSICS_DT, serializeGameState, InputState } from '@open-haxball/shared';
import { ServerMessage, serializeServerMessage } from './protocol.js';

interface Client {
  ws: WebSocket;
  playerId: string | null;
}

export class Room {
  private game: Game;
  private clients: Map<WebSocket, Client> = new Map();
  private gameLoopInterval: NodeJS.Timeout | null = null;
  private nextPlayerId = 1;

  constructor() {
    this.game = new Game();
  }

  start(): void {
    if (this.gameLoopInterval) return;

    this.gameLoopInterval = setInterval(() => {
      this.tick();
    }, PHYSICS_DT * 1000);

    console.log('Game loop started');
  }

  stop(): void {
    if (this.gameLoopInterval) {
      clearInterval(this.gameLoopInterval);
      this.gameLoopInterval = null;
    }
    console.log('Game loop stopped');
  }

  addClient(ws: WebSocket): void {
    this.clients.set(ws, { ws, playerId: null });
    console.log(`Client connected. Total clients: ${this.clients.size}`);
  }

  removeClient(ws: WebSocket): void {
    const client = this.clients.get(ws);
    if (client?.playerId) {
      this.game.removePlayer(client.playerId);
      this.broadcast({
        type: 'playerLeft',
        data: { playerId: client.playerId },
      });
      console.log(`Player ${client.playerId} left`);
    }
    this.clients.delete(ws);
    console.log(`Client disconnected. Total clients: ${this.clients.size}`);
  }

  handleJoin(ws: WebSocket, name: string): void {
    const client = this.clients.get(ws);
    if (!client) return;

    if (client.playerId) {
      this.send(ws, { type: 'error', data: { message: 'Already joined' } });
      return;
    }

    const playerId = `player_${this.nextPlayerId++}`;
    const team = this.game.getAutoTeam();
    const player = this.game.addPlayer(playerId, name, team);

    client.playerId = playerId;

    // Send join confirmation to the player
    this.send(ws, {
      type: 'joined',
      data: { playerId, team },
    });

    // Broadcast to all other clients
    this.broadcast(
      {
        type: 'playerJoined',
        data: { playerId, name, team },
      },
      ws
    );

    console.log(`Player ${name} (${playerId}) joined team ${team}`);
  }

  handleInput(ws: WebSocket, input: InputState): void {
    const client = this.clients.get(ws);
    if (!client?.playerId) return;

    this.game.setPlayerInput(client.playerId, input);
  }

  private tick(): void {
    this.game.update(PHYSICS_DT);

    // Broadcast state to all clients
    const state = serializeGameState(this.game.state);
    this.broadcast({ type: 'state', data: state });
  }

  private send(ws: WebSocket, message: ServerMessage): void {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(serializeServerMessage(message));
    }
  }

  private broadcast(message: ServerMessage, exclude?: WebSocket): void {
    const data = serializeServerMessage(message);
    for (const [ws] of this.clients) {
      if (ws !== exclude && ws.readyState === WebSocket.OPEN) {
        ws.send(data);
      }
    }
  }
}

File: packages/server-rust/src/room.rs
use std::collections::HashMap;
use std::time::Instant;

use tokio::sync::mpsc::UnboundedSender;

use crate::game::Game;
use crate::physics::world::PHYSICS_DT;
use crate::protocol::{ClientMessage, InputState, ServerMessage, Team};

// Kick players after 5 minutes of no input
const IDLE_TIMEOUT_SECS: u64 = 300;

pub struct Client {
    pub sender: UnboundedSender<String>,
    pub player_id: Option<String>,
    pub last_activity: Instant,
    pub join_timestamp: Instant,
}

pub struct Room {
    pub game: Game,
    pub clients: HashMap<usize, Client>,
    next_player_id: u32,
    next_client_id: usize,
}

impl Room {
    pub fn new() -> Self {
        Self {
            game: Game::new(),
            clients: HashMap::new(),
            next_player_id: 1,
            next_client_id: 1,
        }
    }

    pub fn add_client(&mut self, sender: UnboundedSender<String>) -> usize {
        let client_id = self.next_client_id;
        self.next_client_id += 1;

        let now = Instant::now();
        self.clients.insert(
            client_id,
            Client {
                sender,
                player_id: None,
                last_activity: now,
                join_timestamp: now,
            },
        );

        println!(
            "Client {} connected. Total clients: {}",
            client_id,
            self.clients.len()
        );
        client_id
    }

    pub fn remove_client(&mut self, client_id: usize) {
        if let Some(client) = self.clients.remove(&client_id) {
            if let Some(player_id) = &client.player_id {
                self.game.remove_player(player_id);
                self.broadcast(
                    ServerMessage::PlayerLeft {
                        player_id: player_id.clone(),
                    },
                    Some(client_id),
                );
                println!("Player {} left", player_id);
            }
        }
        println!(
            "Client {} disconnected. Total clients: {}",
            client_id,
            self.clients.len()
        );
    }

    pub fn handle_message(&mut self, client_id: usize, message: ClientMessage) {
        match message {
            ClientMessage::Join { name } => self.handle_join(client_id, name),
            ClientMessage::Input(input) => self.handle_input(client_id, input),
            ClientMessage::Chat { text } => self.handle_chat(client_id, text),
            ClientMessage::SwitchTeam { team } => self.handle_switch_team(client_id, team),
            ClientMessage::RestartMatch => self.handle_restart_match(client_id),
        }
    }

    fn handle_join(&mut self, client_id: usize, name: String) {
        let client = match self.clients.get_mut(&client_id) {
            Some(c) => c,
            None => return,
        };

        if client.player_id.is_some() {
            self.send(
                client_id,
                ServerMessage::Error {
                    message: "Already joined".to_string(),
                },
            );
            return;
        }

        let player_id = format!("player_{}", self.next_player_id);
        self.next_player_id += 1;

        let team = self.game.get_auto_team();
        self.game.add_player(player_id.clone(), name.clone(), team);

        // Update client
        if let Some(client) = self.clients.get_mut(&client_id) {
            client.player_id = Some(player_id.clone());
        }

        // Send join confirmation
        self.send(
            client_id,
            ServerMessage::Joined {
                player_id: player_id.clone(),
                team,
            },
        );

        // Broadcast to others
        self.broadcast(
            ServerMessage::PlayerJoined {
                player_id: player_id.clone(),
                name: name.clone(),
                team,
            },
            Some(client_id),
        );

        println!("Player {} ({}) joined team {:?}", name, player_id, team);
    }

    fn handle_input(&mut self, client_id: usize, input: InputState) {
        let player_id = match self.clients.get_mut(&client_id) {
            Some(c) => {
                c.last_activity = Instant::now();
                c.player_id.clone()
            }
            None => return,
        };

        if let Some(player_id) = player_id {
            self.game.set_player_input(&player_id, input);
        }
    }

    fn handle_chat(&mut self, client_id: usize, text: String) {
        // Update last activity
        let player_id = match self.clients.get_mut(&client_id) {
            Some(c) => {
                c.last_activity = Instant::now();
                c.player_id.clone()
            }
            None => return,
        };

        // Look up player info
        let (player_id, name) = match player_id {
            Some(pid) => match self.game.players.get(&pid) {
                Some(player) => (pid, player.name.clone()),
                None => return,
            },
            None => return,
        };

        // Truncate text to 200 chars
        let text: String = text.chars().take(200).collect();

        // Broadcast to all clients
        self.broadcast(
            ServerMessage::Chat {
                player_id,
                name,
                text,
            },
            None,
        );
    }

    fn handle_switch_team(&mut self, client_id: usize, team: Team) {
        let player_id = match self.clients.get_mut(&client_id) {
            Some(c) => {
                c.last_activity = Instant::now();
                c.player_id.clone()
            }
            None => return,
        };

        if let Some(pid) = player_id {
            self.game.switch_player_team(&pid, team);
        }
    }

    fn handle_restart_match(&mut self, client_id: usize) {
        // Update last activity and get player ID
        let player_id = match self.clients.get_mut(&client_id) {
            Some(c) => {
                c.last_activity = Instant::now();
                c.player_id.clone()
            }
            None => return,
        };

        // Verify player is host
        let host_id = self.get_host_player_id();
        if player_id.as_ref() != host_id.as_ref() {
            self.send(
                client_id,
                ServerMessage::Error {
                    message: "Only the host can restart the match".to_string(),
                },
            );
            return;
        }

        self.game.restart_match();
    }

    fn get_host_player_id(&self) -> Option<String> {
        self.clients
            .values()
            .filter(|c| c.player_id.is_some())
            .min_by_key(|c| c.join_timestamp)
            .and_then(|c| c.player_id.clone())
    }

    pub fn tick(&mut self) {
        // Check for idle players and collect IDs to remove
        let idle_clients: Vec<usize> = self
            .clients
            .iter()
            .filter(|(_, client)| {
                client.player_id.is_some()
                    && client.last_activity.elapsed().as_secs() > IDLE_TIMEOUT_SECS
            })
            .map(|(id, _)| *id)
            .collect();

        // Kick idle players
        for client_id in idle_clients {
            println!("Kicking idle client {}", client_id);
            self.remove_client(client_id);
        }

        self.game.update(PHYSICS_DT);

        // Get host player ID
        let host_id = self.get_host_player_id();

        // Send state to each client with their is_host status
        for (client_id, client) in &self.clients {
            let is_host =
                client.player_id.as_ref() == host_id.as_ref() && client.player_id.is_some();
            let state = self.game.serialize_state(is_host);
            self.send(*client_id, ServerMessage::State(state));
        }
    }

    fn send(&self, client_id: usize, message: ServerMessage) {
        if let Some(client) = self.clients.get(&client_id) {
            if let Ok(json) = serde_json::to_string(&message) {
                let _ = client.sender.send(json);
            }
        }
    }

    fn broadcast(&self, message: ServerMessage, exclude: Option<usize>) {
        if let Ok(json) = serde_json::to_string(&message) {
            for (id, client) in &self.clients {
                if Some(*id) != exclude {
                    let _ = client.sender.send(json.clone());
                }
            }
        }
    }
}

impl Default for Room {
    fn default() -> Self {
        Self::new()
    }
}
